### PR TITLE
Tag Expression Language on Alerting

### PR DESCRIPTION
**Pull request by lucasponce**

### CODE REVIEW

**Comment by lucasponce** [comment]
This PR is still in WIP.
Integration tests are not yet introduced.
But I have created it for an earlier review with @stefannegrea as it might share some features with metrics side.

Strategy of this parser is to optimize expression providing shortcuts, most of work on the parser orders the tree with "heavy" operations at the end to shorcut the expression as soon as an empty set is identified.

As commented, still progressing on it, but feel free to comment or suggest some aspect.

**Comment by lucasponce** [comment]
Tests added for Alerts on ITest (basically to validate the REST parameters) and PersistenceTest (to perform a full coverage of the feature).

**Comment by lucasponce** [comment]
Tags expression support added on alerts/events API.
Feature itself it is almost complete.
Pending to sync with latest changes on language expression defined on metrics.

**Comment by lucasponce** [comment]
Moving to NIN to NOTIN creates some conflicts when we introduce the NOT operator.
So, I think the clean solution is to separate NOTIN -> NOT IN.
Then we might have:
tagA and (not tagB or tagC not in ['a','b','c'])
It happens that NOTIN is formed by two existing tokens (NOT and IN) and processing all of them implies to refactor the current rules.
I guess the space is a good workaround and it doesn't affect into the clarity of the language.
cc @stefannegrea

**Comment by lucasponce** [comment]
Please @jshaughn can you start to take a look as well ?
It might need some minor refinement, but feature is almost ready for review.

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/CassAlertsServiceImpl.java [hunk 1]
```diff
@@ -1142,6 +1323,23 @@ private void filterByStatuses(String tenantId, AlertsCriteria criteria, Collecti
                 }
 
                 /*
+                    Get eventIds via tagQuery
+                 */
+                if (criteria.hasTagQueryCriteria()) {
+                    Set<String> alertIdsFilteredByTagQuery = getIdsByTagQuery(tenantId, TagType.EVENT,
```
Comment (jshaughn) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/CassAlertsServiceImpl.java]: alertIdsFilteredByTagQuery -> eventIdsFilteredByTagQuery

File: hawkular-alerts-engine/src/test/java/org/hawkular/alerts/engine/tags/ExpressionTagQueryParserTest.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            // Expected
+        }
+    }
+
+    @Test
+    public void t01_twoTagExpressions() throws Exception {
+        String e1 = "tagA and not tagB";
+        assertEquals("and(tagA, not tagB)", parser.parse(e1));
+
+        // not tag is heavy, so tagA is prioritized to shortcut the analysis if tagA is empty
```
Comment (jshaughn) [inline hawkular-alerts-engine/src/test/java/org/hawkular/alerts/engine/tags/ExpressionTagQueryParserTest.java]: good.

File: hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/AlertsHandler.java [hunk 3]
```diff
@@ -115,11 +115,15 @@ public Response findAlerts(
...
             @QueryParam("severities")
             final String severities,
-            @ApiParam(required = false, value = "Filter out events for unspecified tags.",
+            @ApiParam(required = false, value = "Filter out alerts for unspecified tags.",
                 allowableValues = "Comma separated list of tags, each tag of format 'name\\|value'. + \n" +
                         "Specify '*' for value to match all values.")
             @QueryParam("tags")
             final String tags,
+            @ApiParam(required = false, value = "Filter out alerts for unspecified tags.",
+                allowableValues = "A tag query expression.")
```
Comment (jshaughn) [inline hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/AlertsHandler.java]: Can we add any hint as to what "a tag query expression" is , a doc pointer or how to define it?

File: hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/CrossTenantHandler.java [hunk 4]
```diff
@@ -114,6 +114,10 @@ public Response findAlerts(
                             "Specify '*' for value to match all values.")
             @QueryParam("tags")
             final String tags,
+            @ApiParam(required = false, value = "Filter out alerts for unspecified tags.",
+                    allowableValues = "A tag query expression.")
+            @QueryParam("tagExp")
+            final String tagExp,
```
Comment (jshaughn) [inline hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/CrossTenantHandler.java]: for consistency, probably tagExp -> tagQuery

File: hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/CrossTenantHandler.java [hunk 5]
```diff
@@ -196,6 +200,10 @@ public Response findEvents(
                             "Specify '*' for value to match all values.")
             @QueryParam("tags")
             final String tags,
+            @ApiParam(required = false, value = "Filter out events for unspecified tags.",
+                    allowableValues = "A tag query expression.")
+            @QueryParam("tagExp")
+            final String tagExp,
```
Comment (jshaughn) [inline hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/CrossTenantHandler.java]: same, for consitency, probably tagExp -> tagQuery

File: hawkular-alerts-rest-tests/src/test/groovy/org/hawkular/alerts/rest/EventsITest.groovy [hunk 6]
```diff
@@ -58,6 +58,9 @@ class EventsITest extends AbstractITestBase {
 
         resp = client.get(path: "events", query: [tags:"tag-01|value-01,tag-02|value-02",thin:true] )
         assertEquals(200, resp.status)
+
+        resp = client.get(path: "events", query: [tagQuery:"tagA or (tagB and tagC in ['e.*', 'f.*'])"] )
```
Comment (jshaughn) [inline hawkular-alerts-rest-tests/src/test/groovy/org/hawkular/alerts/rest/EventsITest.groovy]: A naive question but I guess when you see something like "tagA" it just means the tag with name "tagA" and any value?  So in a lot of the tests here we're just doing matching against a name and any value?  Am I right that the test above logically expands to:

name=tagA:value=* or ( name=tagB:value=* and ( name=tagC:value=e.* or name=tagC:value=f.* ))

File: hawkular-alerts-api/src/main/java/org/hawkular/alerts/api/services/AlertsCriteria.java [hunk 7]
```diff
@@ -47,15 +47,16 @@
     String triggerId = null;
     Collection<String> triggerIds = null;
     Map<String, String> tags = null;
+    String tagQuery = null;
```
Comment (lucasponce) [inline hawkular-alerts-api/src/main/java/org/hawkular/alerts/api/services/AlertsCriteria.java]: Good point.
We certainly we could do that.
I guess we could maintain the old tags query map and convert it into a tagQuery.
The implementation combines mostly of the two approaches, so I don't see issues doing that.
I will take a look at it.

File: hawkular-alerts-api/src/main/java/org/hawkular/alerts/api/services/AlertsCriteria.java [hunk 8]
```diff
@@ -248,6 +250,28 @@ public void addTag(String name, String value) {
...
+     *        <expression> ::= ( <tag_key> | <not> <tag_key> | <tag_key> <boolean_operator> <tag_value> |
+     *               <tag_key> <array_operator> <array> )
+     *        <not> ::= [ "NOT" | "not" ]
+     *        <logical_operator> ::= [ "AND" | "OR" | "and" | "or" ]
+     *        <boolean_operator> ::= [ "==" | "!=" ]
+     *        <array_operator> ::= [ "IN" | "NOT IN" | "in" | "not in" ]
+     *        <array> ::= ( "[" "]" | "[" ( "," <tag_value> )* )
+     *        <tag_key> ::= <identifier>                                // Tag identifier
+     *        <tag_value> ::= ( "'" <regexp> "'" | <simple_value> )     // Regular expression used with quotes
+     */
```
Comment (lucasponce) [inline hawkular-alerts-api/src/main/java/org/hawkular/alerts/api/services/AlertsCriteria.java]: +1
I don't have a preference, either. I introduced "key" nomenclature when I imported the metrics g4.
"name" is fine I will sync with that.

File: hawkular-alerts-api/src/main/java/org/hawkular/alerts/api/services/EventsCriteria.java [hunk 9]
```diff
@@ -169,6 +171,28 @@ public void addTag(String name, String value) {
...
+     *        <expression> ::= ( <tag_key> | <not> <tag_key> | <tag_key> <boolean_operator> <tag_value> |
+     *               <tag_key> <array_operator> <array> )
+     *        <not> ::= [ "NOT" | "not" ]
+     *        <logical_operator> ::= [ "AND" | "OR" | "and" | "or" ]
+     *        <boolean_operator> ::= [ "==" | "!=" ]
+     *        <array_operator> ::= [ "IN" | "NOT IN" | "in" | "not in" ]
+     *        <array> ::= ( "[" "]" | "[" ( "," <tag_value> )* )
+     *        <tag_key> ::= <identifier>                                // Tag identifier
+     *        <tag_value> ::= ( "'" <regexp> "'" | <simple_value> )     // Regular expression used with quotes
+     */
```
Comment (lucasponce) [inline hawkular-alerts-api/src/main/java/org/hawkular/alerts/api/services/EventsCriteria.java]: +1

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/CassAlertsServiceImpl.java [hunk 10]
```diff
@@ -639,6 +645,23 @@ public Event getEvent(String tenantId, String eventId, boolean thin) throws Exce
                 }
 
                 /*
+                    Get alertsIds via tagQuery
```
Comment (lucasponce) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/CassAlertsServiceImpl.java]: Plurals, plurals :-)
+1

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/CassAlertsServiceImpl.java (hunk 1 above)
Comment (lucasponce) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/CassAlertsServiceImpl.java]: +1

File: hawkular-alerts-rest-tests/src/test/groovy/org/hawkular/alerts/rest/EventsITest.groovy (hunk 6 above)
Comment (lucasponce) [inline hawkular-alerts-rest-tests/src/test/groovy/org/hawkular/alerts/rest/EventsITest.groovy]: That's correct.

```
tagA or (tagB and tagC in ['e.*', 'f.*'])
name=tagA:value=* or ( name=tagB:value=* and ( name=tagC:value=e.* or name=tagC:value=f.* ))
```
The below expression are equivalent.

File: hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/AlertsHandler.java (hunk 3 above)
Comment (lucasponce) [inline hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/AlertsHandler.java]: +1
Yes, I put a temporal description here but not filled.
I will try to summarize the BNF grammar (at least an example) and see how it fits in the asciidoc render.

File: hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/CrossTenantHandler.java (hunk 4 above)
Comment (lucasponce) [inline hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/CrossTenantHandler.java]: +1
Yes, this is a bug.

File: hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/CrossTenantHandler.java (hunk 5 above)
Comment (lucasponce) [inline hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/CrossTenantHandler.java]: +1

**Comment by lucasponce** [comment]
Thanks for the review @jshaughn. I will prepare a commit with the changes.

**Comment by lucasponce** [comment]
@jshaughn I've sent a commit with the suggestions and corrections.
Please, let me know if you see anything further or if it is almost ready.
Then, I can squash and prepare a commit with the JIRA.
Thanks.

File: hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/CrossTenantHandler.java [hunk 11]
```diff
@@ -116,8 +137,8 @@ public Response findAlerts(
             final String tags,
             @ApiParam(required = false, value = "Filter out alerts for unspecified tags.",
                     allowableValues = "A tag query expression.")
-            @QueryParam("tagExp")
-            final String tagExp,
+            @QueryParam("tagQuery")
+            final String tagQuery,
```
Comment (jshaughn) [inline hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/CrossTenantHandler.java]: Did you want to use  "[DEPRECATED]" on the original "tags" field?

File: hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/CrossTenantHandler.java [hunk 12]
```diff
@@ -202,8 +252,8 @@ public Response findEvents(
             final String tags,
             @ApiParam(required = false, value = "Filter out events for unspecified tags.",
                     allowableValues = "A tag query expression.")
-            @QueryParam("tagExp")
-            final String tagExp,
+            @QueryParam("tagQuery")
+            final String tagQuery,
```
Comment (jshaughn) [inline hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/CrossTenantHandler.java]: Again, did you want to use  "[DEPRECATED]" on the original "tags" field?

File: hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/EventsHandler.java [hunk 13]
```diff
@@ -249,7 +253,26 @@ public Response deleteTags(
...
+                    "<logical_operator> ::= [ \"AND\" | \"OR\" | \"and\" | \"or\" ] \n" +
+                    "<boolean_operator> ::= [ \"==\" | \"!=\" ] \n" +
+                    "<array_operator> ::= [ \"IN\" | \"NOT IN\" | \"in\" | \"not in\" ] \n" +
+                    "<array> ::= ( \"[\" \"]\" | \"[\" ( \",\" <tag_value> )* ) \n" +
+                    "<tag_name> ::= <identifier> \n" +
+                    "<tag_value> ::= ( \"'\" <regexp> \"'\" | <simple_value> ) \n" +
+                    "; \n" +
+                    "; <identifier> and <simple_value> follow pattern [a-zA-Z_0-9][\\-a-zA-Z_0-9]* \n" +
+                    "; <regexp> follows any valid Java Regular Expression format \n" +
+                    "---- \n",
```
Comment (jshaughn) [inline hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/EventsHandler.java]: And here, Did you want to use  "[DEPRECATED]" on the original "tags" field?

File: hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/EventsHandler.java [hunk 14]
```diff
@@ -347,7 +380,26 @@ public Response deleteEvent(
...
+                    "<logical_operator> ::= [ \"AND\" | \"OR\" | \"and\" | \"or\" ] \n" +
+                    "<boolean_operator> ::= [ \"==\" | \"!=\" ] \n" +
+                    "<array_operator> ::= [ \"IN\" | \"NOT IN\" | \"in\" | \"not in\" ] \n" +
+                    "<array> ::= ( \"[\" \"]\" | \"[\" ( \",\" <tag_value> )* ) +\n" +
+                    "<tag_name> ::= <identifier> \n" +
+                    "<tag_value> ::= ( \"'\" <regexp> \"'\" | <simple_value> ) \n" +
+                    "; \n " +
+                    "; <identifier> and <simple_value> follow pattern [a-zA-Z_0-9][\\-a-zA-Z_0-9]* \n" +
+                    "; <regexp> follows any valid Java Regular Expression format \n" +
+                    "---- \n",
```
Comment (jshaughn) [inline hawkular-alerts-rest/hawkular-alerts-rest-api/src/main/java/org/hawkular/alerts/rest/EventsHandler.java]: One more, Did you want to use  "[DEPRECATED]" on the original "tags" field?

**Comment by lucasponce** [comment]
I squashed the WIP commits and remove the label.
I think is ready to merge, but let me know if there is any additional tasks to finish.
Thanks !

**Pull request merged**